keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -871,32 +880,314 @@ not added to the pod for a long enough time (for example 2 minutes).
-However, as disabling taint-manager is deprecated it is not a concern for this
-KEP.
-
-For Alpha, we implemented a fix for this issue by setting the pod phase as `Failed`
-in podgc.
-
-For Beta, we are going to simplify the code in job controller responsible for
-detecting pod failures. For this purpose we need to make sure that pods stuck
-in the pending phase which are terminating (with set deletionTimestamp) are
-eventually marked as failed. Such pods may either be not scheduled
-(for example when they are unschedulable due to node affinity configuration for
-which no node exists in the cluster) or scheduled (for example when they fail
-to pull the the docker image). In case of non-scheduled pods they will be moved
-to the `Failed` phase by the PodGC controller as it marks as failed all
-non-scheduled terminating pods (after the Alpha changes for this feature, see
-above). We plan to extend the PodGC controller to also move to the failed phase
-(and delete) scheduled pods in the pending state which are terminating.
+When matching a failed pod against Job pod failure policy it is important that
+the pod is actually in the terminal phase (`Failed`), to ensure their state is
+not modified while Job controller matches them against the pod failure policy.
+
+However, there are scenarios in which a pod gets stuck in a non-terminal phase,
+but is doomed to be failed, as it is terminating (has `deletionTimestamp` set).
+In order to workaround this issue, Job controller, when pod failure policy is
+disabled, considers any terminating pod that is in a non-terminal phase as failed.
+Note that, it is important that when Job controller considers such pods as failed
+so that it removes their finalizers and thus allows the API server to complete
+their deletion.
+
+In order to ensure consistency in behavior when handling pod failures when
+pod failure policy is used or not, we need to make sure that all pods which
+are terminating (were previously considered as failed by the Job controller),
+are transitioned to the failed state eventually.
+
+Thus, we review scenarios in which a pod is considered by Job controller as
+failed, but may get stuck in a non-terminal. Note that, the following scenarios
+are not only problematic to the Job controller (due to its use of finalizers),
+but they can be considered as bugs in their own right, as every pod should end
+up in a terminal phase, even if not started
+(see [discussion](https://github.com/kubernetes/enhancements/pull/3757#discussion_r1095197982)).
+
+1. **Orphan pods** (solved in Alpha)
+
+This pod state is characterized by the following:
+- scheduled (the Pod's `.spec.nodeName` is set), but the node no longer exists
+- `Running` or `Pending` phase
+
+Note that, as PodGC sends DELETE request for such pod (DELETE request was sent
+prior to this KEP, but without setting the phase to `Failed`), it becomes also
+terminating (has `deletionTimestamp` set).
```

Comment:
nit: I would avoid using the term "terminating", as for the pod to be terminating in the traditional sense, the pod has have had started. If the pod never started, it may not even be terminating. 

I think the term we should use here is "deleting". See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/object-lifecycle.md#states here it is defined. (and you can maybe link to it from the KEP)

Reply:
Yeah, I guess I was influenced by the terminology used in kubectl, there `Terminating` is displayed even if it has not started. 

For example, if I create a pod via the following Job spec (unschedulable):
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: job-unchedulable
spec:
  template:
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: kubernetes.io/os
                operator: In
                values:
                - something
      restartPolicy: Never
      containers:
      - name: job-unchedulable
        image: centos:7
        command: ["bash"]
        args: ["-c", 'echo "Hello world"']
        imagePullPolicy: IfNotPresent
  backoffLimit: 0
```
Then, do `kubectl delete`, it shows up as `Terminating` in the console for `kubectl get pods`. 

Also, the `Terminating` word is used in PodGC code (prior to this work): https://github.com/kubernetes/kubernetes/blob/8e20effbb7b99c9c5caad4b8f947ca645bbaf21b/pkg/controller/podgc/gc_controller.go#L133.

Reply:
For now, I reference the link and add a note that this is an alternative name. WDYT?

Reply:
SGTM for now.

Reply:
Sounds good, thanks for adding the reference. 

Just as fyi, the term terminating in kubectl comes from [here](https://github.com/kubernetes/kubernetes/blob/0f37b3120643580f632ca12d3e174e7ec447948c/pkg/printers/internalversion/printers.go#L853-L855). `Terminating` shows up if deletion timestamp is set.

My comment is just that in my mind there is distinction between deleting and terminating. `Deleting` == deletion timestamp is set, terminating == the kubelet is terminating the pod, i.e. terminating containers, cleaning up resources etc. As described in the KEP, in any case the pod is `Deleting`, the phase should be updated to terminal.